File: kios_cpp/src/node/test_node.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"

#include "rcl_interfaces/srv/set_parameters_atomically.hpp"
#include "rcl_interfaces/srv/get_parameters.hpp"
#include "rcl_interfaces/msg/parameter.hpp"

#include "kios_interface/msg/mios_state.hpp"
#include "kios_interface/msg/task_state.hpp"
#include "kios_interface/msg/sensor_state.hpp"

#include "kios_utils/kios_utils.hpp"

#include <condition_variable>
#include <mutex>
#include <thread>
#include <chrono>
#include <optional>

using std::placeholders::_1;

class TestNode : public rclcpp::Node
{
public:
    TestNode()
        : Node("test_node"),
          context_clerk_()

    {
        this->declare_parameter("power", true);

        timer_callback_group_ = this->create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

        //* initialize timer
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(&TestNode::timer_callback, this),
            timer_callback_group_);
    }

    bool check_power()
    {
        return this->get_parameter("power").as_bool();
    }

private:
    // callback group
    rclcpp::CallbackGroup::SharedPtr timer_callback_group_;
    kios::DefaultActionContext memory_test;
    kios::ContextClerk context_clerk_;

    // callbacks
    rclcpp::TimerBase::SharedPtr timer_;

    void timer_callback()
    {
        RCLCPP_INFO(this->get_logger(), "Test Loop.");
        auto str = kios::action_phase_to_str(kios::ActionPhase::INITIALIZATION);
        RCLCPP_INFO_STREAM(this->get_logger(), "the string is " << str.value() << " ! ");
        RCLCPP_INFO_STREAM(this->get_logger(), "the value is " << memory_test.get_default_context(kios::ActionPhase::CARTESIAN_MOVE).value() << " ! ");
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);

    auto test_node = std::make_shared<TestNode>();
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(test_node);

    executor.spin();

    rclcpp::shutdown();
    return 0;
}

File: kios_py/kios_py/test_node.py
import rclpy
from rclpy.node import Node
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup, ReentrantCallbackGroup

import time
from datetime import datetime


class TestNode(Node):

    def __init__(self):
        super().__init__('test_node')

        # declare parameters
        self.declare_parameter('power', True)

        

        timer_callback_group = ReentrantCallbackGroup()

        self.timer = self.create_timer(
            0.5,  # sec
            self.timer_callback,
            callback_group=timer_callback_group)

    def timer_callback(self):
        if self.check_power():
            self.get_logger().info('Power on, timer hit ...')
        else:
            self.get_logger().error('Power off, timer pass ...')
            pass

    def check_power(self):
        if self.has_parameter('power'):
            self.get_logger().error('CHECK POWER')
            return self.get_parameter('power').get_parameter_value().bool_value
        else:
            self.get_logger().error('PARAM MISSING: POWER!')
            return False

    def switch_power(self, turn_on: bool):
        power = rclpy.parameter.Parameter(
            'power',
            rclpy.Parameter.Type.BOOL,
            turn_on
        )
        all_new_parameters = [power]
        self.set_parameters(all_new_parameters)


def main(args=None):
    rclpy.init(args=args)

    test_node = TestNode()

    executor = MultiThreadedExecutor()
    executor.add_node(test_node)

    executor.spin()

    test_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
